pom.xml

```diff
@@ -46,14 +48,6 @@
   </developers>
 
   <dependencies>
-    <dependency>
-      <groupId>io.jenkins.plugins</groupId>
-      <artifactId>commons-lang3-api</artifactId>
-    </dependency>
-    <dependency>
-      <groupId>io.jenkins.plugins</groupId>
-      <artifactId>commons-text-api</artifactId>
-    </dependency>
```

Comment:
These dependencies are used by all downstream plugins. We cannot remove them without breaking the downstream plugins.

Reply:
> > We cannot remove them without breaking the downstream plugins.
> 
> if they are used by downstream plugins - then it would be a bug in those plugins to not declare a dependency on these plugins in their pom, they need to be fixed, but that is irrespective of this PR.

Why is this a bug? This is a feature of Maven. Commons lang (and some other dependencies) are provided by this plugin-util library so that you do not need to explicitly need to declare those dependencies in your plugins. So there is only a single place to change those dependencies.

Reply:
> Why is this a bug? This is a feature of Maven

the feature I think you are talking about is [transitive dependencies](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#transitive-dependencies).

The maven feature will give you transitive dependencies that 3rd parties use in your classpath, this is so when compliling/testing *your* code it will work when run with the library that needs that code.

> Maven avoids the need to discover and specify the libraries **that your own dependencies require** by including transitive dependencies automatically.  (emphasis mine)

However **if** you also use those dependencies directly you are only getting the dependency via the 3rd party thus if the 3rd party changes you will end up with a broken build because you are using the library and not declaring it, and this is then incorrect.

To see this in action (and why it is incorrect) run [`mvn dependency:analyze`](https://maven.apache.org/plugins/maven-dependency-plugin/analyze-mojo.html) and pay attention to the warning "used but not declared".

If you use something you should *always* declare it.

(now dependencies coming from Jenkins itself are somewhat special because of the way Jenkins and the hpi plugin themselves abuse some maven constructs - but for library and plugin dependencies you should never rely on transitive dependencies if you directly use something yourself).

See also https://books.sonatype.com/mvnex-book/reference/optimizing-sect-dependency-plugin.html

Reply:
In general you are right, it is not recommended (while possible) to use transitive dependencies of a library that you are using. 

However in the case of the plugin-util API: this is one of the reasons for the existence for this plugin. This plugin provides libraries with a fixed version number for other plugins (like Jenkins does). It contains these libraries as copied classes or as dependency to another API plugin (before the birth of commons-lang-api-plugin these classes were part of the plugin-util as well). This can be done in a similar way by changing the parent pom but it is done via transitive dependencies for the historical reason I mentioned before.

Reply:
> However in the case of the plugin-util API: this is one of the reasons for the existence for this plugin. This plugin provides libraries with a fixed version number for other plugins (like Jenkins does)

Maven provides a BOM for this purpose, and there exists a [bom implementation for plugins](https://github.com/jenkinsci/bom) that does even more than just give a set of versions - it provides a known good tested set of versions.

>  This can be done in a similar way by changing the parent pom

no need to change any parent pom, as mentioned above.  just depenencyManagement with the bom at `scope` `import` for the plugin who wishes to depend on other plugins with a known version.

Reply:
> > However in the case of the plugin-util API: this is one of the reasons for the existence for this plugin. This plugin provides libraries with a fixed version number for other plugins (like Jenkins does)
> 
> Maven provides a BOM for this purpose, and there exists a [bom implementation for plugins](https://github.com/jenkinsci/bom) that does even more than just give a set of versions - it provides a known good tested set of versions.

The BOM just provides the version, not the library. But this is what I want to achieve, a single entry point to define the dependencies of some fundamental libraries that should be used by all plugins.

Reply:
> The BOM just provides the version, not the library. But this is what I want to achieve, a single entry point to define the dependencies of some fundamental libraries that should be used by all plugins.

then perhaps you should create a new library-api plugin.